In [57]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

# Lesson 01: Machine Learning Foundations and First Model

## Learning Objectives

After completing this lesson, I will be able to:

- Understand Artificial Intelligence, Machine Learning, and Deep Learning
- Understand features, target, and samples
- Differentiate classification and regression
- Understand supervised and unsupervised learning
- Understand training and testing data
- Build my first machine learning model using Scikit-learn

## 1. Artificial Intelligence, Machine Learning, and Deep Learning

### Artificial Intelligence

Artificial Intelligence (AI) is the broad field of creating computer systems that can perform tasks that normally require human intelligence.

Examples:
- Face recognition
- Voice assistants
- Recommendation systems
- Self-driving systems

### Machine Learning

Machine Learning (ML) is a subset of AI where computers learn patterns from data instead of being explicitly programmed with every rule.

### Deep Learning

Deep Learning (DL) is a subset of Machine Learning that mainly uses multi-layer neural networks.

### Relationship

Artificial Intelligence  
↓  
Machine Learning  
↓  
Deep Learning

## 2. Feature, Target, and Sample

### Sample

A sample is one individual observation or one row in a dataset.

Example:

| Age | Salary | Experience | Loan Approved |
|---|---:|---:|---|
| 25 | 30000 | 2 | Yes |
| 35 | 50000 | 8 | No |

Each row is one **sample**.

### Feature

A feature is an input variable that the machine learning model uses to make a prediction.

In the example above:

- Age
- Salary
- Experience

are the **features**.

### Target

The target is the value that we want the machine learning model to predict.

In the example above:

**Loan Approved** is the target.

### Simple Rule

Features = Input to the model  
Target = What the model predicts  
Sample = One row or observation in the dataset

## 3. Classification vs Regression

Machine learning problems can be divided into different types. Two important supervised learning tasks are classification and regression.

### Classification

Classification is used when the target is a category or class.

Examples:

- Email → Spam / Not Spam
- Patient → Disease / Healthy
- File → Malware / Benign
- Flower → Setosa / Versicolor / Virginica

### Regression

Regression is used when the target is a continuous numerical value.

Examples:

- House → Price
- Weather → Temperature
- Product → Sales Amount
- Employee → Salary

### Easy Shortcut

Category → Classification  
Continuous Number → Regression

## 4. Supervised and Unsupervised Learning

### Supervised Learning

In supervised learning, the training dataset contains both input features and the correct target values.

The model learns the relationship between the features and the target.

Example:

Student Information + Pass/Fail Result  
↓  
Train Model  
↓  
Predict Pass/Fail for a New Student

Supervised learning mainly includes:

- Classification
- Regression

### Unsupervised Learning

In unsupervised learning, the dataset does not contain a target label.

The model tries to find hidden patterns or groups in the data.

Example:

Customer Information  
↓  
Machine Learning Algorithm  
↓  
Different Customer Groups

Common unsupervised learning tasks include:

- Clustering
- Dimensionality Reduction

## 5. Loading Our First Machine Learning Dataset

For our first machine learning experiment, we will use the Iris dataset.

The Iris dataset contains measurements of flowers from three different species:

- Setosa
- Versicolor
- Virginica

The dataset contains four input features:

- Sepal Length
- Sepal Width
- Petal Length
- Petal Width

Our target is the flower species.

In [58]:
from sklearn.datasets import load_iris

iris = load_iris()

print(iris)

{'data': array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
     

In [59]:
X = iris.data
y = iris.target

print("X:")
print(X)

print("\ny:")
print(y)

X:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1

In [60]:
print("Shape of X:", X.shape)
print("Shape of y:", y.shape)

Shape of X: (150, 4)
Shape of y: (150,)


In [61]:
print("Feature Names:")
print(iris.feature_names)

print("\nTarget Names:")
print(iris.target_names)

Feature Names:
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

Target Names:
['setosa' 'versicolor' 'virginica']


In [62]:
print("First 5 Samples:")
print(X[:5])

print("\nFirst 5 Targets:")
print(y[:5])

First 5 Samples:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]

First 5 Targets:
[0 0 0 0 0]


In [63]:
iris.target_names[0]

np.str_('setosa')

In [64]:
sample_number = 0

print("Features:", X[sample_number])
print("Target Number:", y[sample_number])
print("Target Name:", iris.target_names[y[sample_number]])

Features: [5.1 3.5 1.4 0.2]
Target Number: 0
Target Name: setosa


### What I Learned

- `X` contains the input features.
- `y` contains the target labels.
- The Iris dataset contains 150 samples.
- Each sample has 4 features.
- The target contains 3 flower classes.
- Target `0` = Setosa
- Target `1` = Versicolor
- Target `2` = Virginica

## 6. Viewing the Dataset with Pandas

Machine learning datasets are easier to understand when we view them as a table.

We will use the Pandas library to convert the Iris dataset into a DataFrame.

A DataFrame contains rows and columns, similar to an Excel table.

In [65]:
import pandas as pd

df = pd.DataFrame(
    X,
    columns=iris.feature_names
)

df["target"] = y

df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


In [66]:
df["species"] = df["target"].map({
    0: "setosa",
    1: "versicolor",
    2: "virginica"
})

df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target,species
0,5.1,3.5,1.4,0.2,0,setosa
1,4.9,3.0,1.4,0.2,0,setosa
2,4.7,3.2,1.3,0.2,0,setosa
3,4.6,3.1,1.5,0.2,0,setosa
4,5.0,3.6,1.4,0.2,0,setosa


In [67]:
print("Dataset Shape:", df.shape)

Dataset Shape: (150, 6)


In [68]:
print(df.columns)

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'target', 'species'],
      dtype='object')


In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
 5   species            150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


In [70]:
print(df.isnull().sum())

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
target               0
species              0
dtype: int64


Basic statistics

In [71]:
df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333,1.000000
std,0.828066,0.435866,1.765298,0.762238,0.819232
min,4.300000,2.000000,1.000000,0.100000,0.000000
25%,5.100000,2.800000,1.600000,0.300000,0.000000
50%,5.800000,3.000000,4.350000,1.300000,1.000000
75%,6.400000,3.300000,5.100000,1.800000,2.000000
max,7.900000,4.400000,6.900000,2.500000,2.000000


Class distribution

In [72]:
print(df["species"].value_counts())

species
setosa        50
versicolor    50
virginica     50
Name: count, dtype: int64


In [73]:
df.head(10)

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target,species
0,5.1,3.5,1.4,0.2,0,setosa
1,4.9,3.0,1.4,0.2,0,setosa
2,4.7,3.2,1.3,0.2,0,setosa
3,4.6,3.1,1.5,0.2,0,setosa
4,5.0,3.6,1.4,0.2,0,setosa
5,5.4,3.9,1.7,0.4,0,setosa
6,4.6,3.4,1.4,0.3,0,setosa
7,5.0,3.4,1.5,0.2,0,setosa
8,4.4,2.9,1.4,0.2,0,setosa
9,4.9,3.1,1.5,0.1,0,setosa


In [74]:
df.tail()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target,species
145,6.7,3.0,5.2,2.3,2,virginica
146,6.3,2.5,5.0,1.9,2,virginica
147,6.5,3.0,5.2,2.0,2,virginica
148,6.2,3.4,5.4,2.3,2,virginica
149,5.9,3.0,5.1,1.8,2,virginica


In [75]:
df.sample()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target,species
114,5.8,2.8,5.1,2.4,2,virginica


### What I Learned

- Pandas DataFrame helps us view data in tabular form.
- `head()` shows the first rows of a dataset.
- `shape` shows the number of rows and columns.
- `info()` gives information about columns and data types.
- `isnull().sum()` checks missing values.
- `describe()` gives basic statistical information.
- `value_counts()` shows how many samples belong to each class.
- The Iris dataset is balanced because each species has 50 samples.

## 7. Train-Test Split

A machine learning model should be evaluated on data that it did not see during training.

Therefore, we divide the dataset into two parts:

- Training Data
- Testing Data

### Training Data
The model learns patterns from the training data.

### Testing Data
The testing data is used to evaluate how well the trained model performs on unseen data.

A common split is:

- 80% Training Data
- 20% Testing Data

In [76]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()

X = iris.data
y = iris.target

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Done")
print(X_train.shape)
print(X_test.shape)

Done
(120, 4)
(30, 4)


In [77]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (120, 4)
X_test shape: (30, 4)
y_train shape: (120,)
y_test shape: (30,)


### Why do we use `random_state=42`?

The train-test split is random.

Using a fixed `random_state` makes the result reproducible.

This means that if we run the code again, we will get the same train-test split.

### Why do we use `stratify=y`?

`stratify=y` keeps the class distribution approximately the same in both training and testing datasets.

This is especially useful for classification problems.

In [78]:
import numpy as np

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

# Load dataset
iris = load_iris()

# Features and target
X = iris.data
y = iris.target

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

# Check shapes
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

# Check class distribution
print("\nTraining class counts:")
print(np.bincount(y_train))

print("\nTesting class counts:")
print(np.bincount(y_test))

X_train shape: (120, 4)
X_test shape: (30, 4)
y_train shape: (120,)
y_test shape: (30,)

Training class counts:
[40 40 40]

Testing class counts:
[10 10 10]


In [79]:
import numpy as np

print("Training class counts:")
print(np.bincount(y_train))

print("\nTesting class counts:")
print(np.bincount(y_test))

Training class counts:
[40 40 40]

Testing class counts:
[10 10 10]


In [80]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(
    X,
    y,
    test_size=0.30,
    random_state=42,
    stratify=y
)

print("Training:", X_train2.shape)
print("Testing:", X_test2.shape)

Training: (105, 4)
Testing: (45, 4)


### What I Learned

- A dataset should be divided into training and testing parts.
- Training data is used to train the model.
- Testing data is used to evaluate unseen-data performance.
- `test_size=0.20` means 20% of the data is used for testing.
- `random_state` makes the split reproducible.
- `stratify=y` preserves the class distribution.

In [81]:
## 8. First Machine Learning Model: Logistic Regression

Now we will train our first machine learning model.

We will use Logistic Regression.

Although its name contains "Regression", Logistic Regression is mainly used for classification problems.

Our task is:

Flower Measurements  
↓  
Logistic Regression Model  
↓  
Predict Flower Species

SyntaxError: invalid character '↓' (U+2193) (1355403801.py, line 12)

In [ ]:
print("Training features:", X_train.shape)
print("Testing features:", X_test.shape)
print("Training targets:", y_train.shape)
print("Testing targets:", y_test.shape)

Model create


In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=300)

print(model)

model traning


In [ ]:
model.fit(X_train, y_train)

print("Model training completed!")

In [ ]:
model.fit(X_train, y_train)

test data predict

In [ ]:
y_pred = model.predict(X_test)

print(y_pred)

In [ ]:
print("Predicted:")
print(y_pred[:10])

print("\nActual:")
print(y_test[:10])

In [ ]:
for i in range(10):
    predicted_name = iris.target_names[y_pred[i]]
    actual_name = iris.target_names[y_test[i]]

    print(
        "Actual:", actual_name,
        "| Predicted:", predicted_name
    )

In [ ]:
print("Sample features:")
print(X_test[0])

print("\nActual class:")
print(iris.target_names[y_test[0]])

print("\nPredicted class:")
print(iris.target_names[y_pred[0]])

In [ ]:
new_flower = [[5.1, 3.5, 1.4, 0.2]]

prediction = model.predict(new_flower)

print("Prediction number:", prediction[0])
print("Predicted species:", iris.target_names[prediction[0]])

### What I Learned

- Logistic Regression can be used for classification.
- `model = LogisticRegression()` creates the model.
- `fit()` trains the model using training data.
- `predict()` makes predictions on new or unseen data.
- `X_train` contains training features.
- `y_train` contains the correct training labels.
- `X_test` contains unseen test features.
- `y_test` contains the actual test labels.
- `y_pred` contains the model's predicted labels.

In [ ]:
## 9. Model Evaluation

After training a machine learning model, we need to evaluate its performance.

For classification problems, some common evaluation metrics are:

- Accuracy
- Confusion Matrix
- Precision
- Recall
- F1-score

In this lesson, we will use these metrics to evaluate our Logistic Regression model.

In [ ]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)
print("Accuracy Percentage:", accuracy * 100, "%")

### Accuracy

Accuracy measures how many predictions are correct out of all predictions.

Formula:

Accuracy = Correct Predictions / Total Predictions

In [ ]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

### Confusion Matrix

A confusion matrix shows how many samples were predicted correctly and incorrectly for each class.

- Rows represent actual classes.
- Columns represent predicted classes.
- Diagonal values represent correct predictions.
- Off-diagonal values represent incorrect predictions.

In [ ]:
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

disp = ConfusionMatrixDisplay(
    confusion_matrix=cm,
    display_labels=iris.target_names
)

disp.plot()
plt.title("Confusion Matrix")
plt.show()

In [ ]:
from sklearn.metrics import classification_report

report = classification_report(
    y_test,
    y_pred,
    target_names=iris.target_names
)

print(report)

### Classification Report

The classification report provides more detailed evaluation metrics.

- **Precision**: How many predicted positive cases were actually correct.
- **Recall**: How many actual positive cases were correctly identified.
- **F1-score**: Balance between precision and recall.
- **Support**: Number of actual samples in each class.

In [ ]:
import pandas as pd

results = pd.DataFrame({
    "Actual": y_test,
    "Predicted": y_pred
})

results["Actual Species"] = results["Actual"].map({
    0: "setosa",
    1: "versicolor",
    2: "virginica"
})

results["Predicted Species"] = results["Predicted"].map({
    0: "setosa",
    1: "versicolor",
    2: "virginica"
})

results.head(10)

In [ ]:
results["Correct"] = results["Actual"] == results["Predicted"]

results.head(10)

In [ ]:
results[results["Correct"] == False]

In [ ]:
correct_predictions = (y_test == y_pred).sum()
wrong_predictions = (y_test != y_pred).sum()

print("Correct Predictions:", correct_predictions)
print("Wrong Predictions:", wrong_predictions)

### What I Learned

- Accuracy shows the overall percentage of correct predictions.
- A confusion matrix shows class-wise correct and incorrect predictions.
- Precision measures how reliable positive predictions are.
- Recall measures how many actual positive samples were correctly detected.
- F1-score balances precision and recall.
- Classification reports provide detailed class-wise model performance.
- Model evaluation is necessary before trusting predictions.

## 10. Predicting New Samples

After training and evaluating the model, we can use it to predict the class of new unseen samples.

For the Iris dataset, each new sample must contain four features in this order:

1. Sepal Length
2. Sepal Width
3. Petal Length
4. Petal Width

In [ ]:
new_flower = [[5.1, 3.5, 1.4, 0.2]]

prediction = model.predict(new_flower)

print("Prediction Number:", prediction[0])
print("Predicted Species:", iris.target_names[prediction[0]])

In [ ]:
new_samples = [
    [5.0, 3.4, 1.5, 0.2],
    [6.0, 2.9, 4.5, 1.5],
    [6.7, 3.1, 5.6, 2.4]
]

predictions = model.predict(new_samples)

print(predictions)

In [ ]:
for i, prediction in enumerate(predictions):
    print(
        "Sample", i + 1,
        "->",
        iris.target_names[prediction]
    )

In [ ]:
probabilities = model.predict_proba(new_samples)

print(probabilities)

In [ ]:
for i, probs in enumerate(probabilities):
    print(f"\nSample {i + 1}")

    for class_name, probability in zip(
        iris.target_names,
        probs
    ):
        print(
            class_name,
            ":",
            round(probability, 4)
        )

In [ ]:
test_flower = [[6.5, 3.0, 5.5, 2.0]]

prediction = model.predict(test_flower)
probability = model.predict_proba(test_flower)

print(
    "Predicted Species:",
    iris.target_names[prediction[0]]
)

print(
    "Probabilities:",
    probability
)

## 11. Practice Exercises

### Exercise 1
Print the first 10 samples from the Iris dataset.

### Exercise 2
Print the last 10 samples from the Pandas DataFrame.

### Exercise 3
Change the test size from 20% to 30% and observe the train and test shapes.

### Exercise 4
Train Logistic Regression again using the new train-test split.

### Exercise 5
Calculate the model accuracy.

### Exercise 6
Display the confusion matrix.

### Exercise 7
Predict the species of at least 3 new flowers.

### Exercise 8
Use `predict_proba()` and observe the probability for each class.

### Exercise 9
Find all incorrect predictions from the test set.

### Exercise 10
Explain in your own words:
- Feature
- Target
- Training data
- Testing data
- `fit()`
- `predict()`
- Accuracy

In [ ]:
print("First 10 samples:")
print(X[:10])

In [ ]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()

X = iris.data
y = iris.target

df = pd.DataFrame(
    X,
    columns=iris.feature_names
)

df["target"] = y

df["species"] = df["target"].map({
    0: "setosa",
    1: "versicolor",
    2: "virginica"
})

print("DataFrame created successfully.")

In [ ]:
df.tail(10)

In [ ]:
from sklearn.model_selection import train_test_split

X_train_30, X_test_30, y_train_30, y_test_30 = train_test_split(
    X,
    y,
    test_size=0.30,
    random_state=42,
    stratify=y
)

print("Training shape:", X_train_30.shape)
print("Testing shape:", X_test_30.shape)

print("Training target shape:", y_train_30.shape)
print("Testing target shape:", y_test_30.shape)

In [ ]:
from sklearn.linear_model import LogisticRegression

model_30 = LogisticRegression(max_iter=300)

model_30.fit(X_train_30, y_train_30)

print("Model training completed.")

In [ ]:
model_30.fit(X_train_30, y_train_30)

In [ ]:
y_pred_30 = model_30.predict(X_test_30)

print(y_pred_30)

In [ ]:
from sklearn.metrics import accuracy_score

accuracy_30 = accuracy_score(
    y_test_30,
    y_pred_30
)

print("Accuracy:", accuracy_30)
print("Accuracy Percentage:", accuracy_30 * 100, "%")

In [ ]:
from sklearn.metrics import confusion_matrix

cm_30 = confusion_matrix(
    y_test_30,
    y_pred_30
)

print("Confusion Matrix:")
print(cm_30)

In [ ]:
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

display = ConfusionMatrixDisplay(
    confusion_matrix=cm_30,
    display_labels=iris.target_names
)

display.plot()

plt.title("Confusion Matrix")
plt.show()

In [ ]:
new_flowers = [
    [5.0, 3.4, 1.5, 0.2],
    [6.0, 2.9, 4.5, 1.5],
    [6.7, 3.1, 5.6, 2.4]
]

predictions = model_30.predict(new_flowers)

for i, prediction in enumerate(predictions):
    print(
        "Flower",
        i + 1,
        "->",
        iris.target_names[prediction]
    )

In [ ]:
import pandas as pd

results_30 = pd.DataFrame({
    "Actual": y_test_30,
    "Predicted": y_pred_30
})

results_30.head()

In [ ]:
probabilities = model_30.predict_proba(new_flowers)

print(probabilities)

In [ ]:
import pandas as pd

results_30 = pd.DataFrame({
    "Actual": y_test_30,
    "Predicted": y_pred_30
})

results_30.head()

In [ ]:
results_30["Actual Species"] = [
    iris.target_names[x]
    for x in y_test_30
]

results_30["Predicted Species"] = [
    iris.target_names[x]
    for x in y_pred_30
]

results_30["Correct"] = (
    results_30["Actual"]
    ==
    results_30["Predicted"]
)

results_30.head()

In [ ]:
incorrect_predictions = results_30[
    results_30["Correct"] == False
]

incorrect_predictions

In [ ]:
print(
    "Total incorrect predictions:",
    len(incorrect_predictions)
)

In [ ]:
print(
    "Wrong Predictions:",
    (y_test_30 != y_pred_30).sum()
)

## Exercise 10: Important Concepts

### Feature

A feature is an input variable used by a machine learning model to make predictions.

Example:

In the Iris dataset:

- Sepal Length
- Sepal Width
- Petal Length
- Petal Width

are features.

### Target

The target is the output that we want the machine learning model to predict.

In the Iris dataset, the flower species is the target.

### Training Data

Training data is the portion of the dataset used by the model to learn patterns.

The model learns from:

`X_train` = Training Features  
`y_train` = Correct Training Labels

### Testing Data

Testing data is unseen data used to evaluate the trained model.

`X_test` contains testing features.

`y_test` contains the actual answers.

### fit()

`fit()` is used to train the machine learning model.

Example:

model.fit(X_train, y_train)

The model learns the relationship between the features and target.

### predict()

`predict()` is used to make predictions using a trained model.

Example:

y_pred = model.predict(X_test)

### Accuracy

Accuracy measures the proportion of predictions that are correct.

Accuracy = Correct Predictions / Total Predictions

For example, if the model correctly predicts 29 out of 30 samples:

Accuracy = 29 / 30 = 96.67%

Dataset
   ↓
Features (X) + Target (y)
 ↓
Train-Test Split
 ↓
X_train, X_test, y_train, y_test
   ↓
Create Model
   ↓
model.fit(X_train, y_train)
 ↓
Trained Model
 ↓
model.predict(X_test)
   ↓
y_pred
 ↓
Compare with y_test
 ↓
Model Evaluation
   ↓
Accuracy
Confusion Matrix
Classification Report